Implementing Cert Picker and PIN Dialogs for Smartcard CBA

melissaahn opened the pull request:
### Note
New smartcard CBA features will be merged into the feature branch `melissaahn/smartcardCBAFeature` until the feature is ready for public previews.
## The Feature
This PR builds on [my first PR](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1729) by using the established connection with the YubiKey device to start the process of smartcard CBA. The parts of the smartcard CBA feature that are being address in this PR are (with corresponding PBIs linked):

- Recognizing that a YubiKey is plugged in, and thus, the user wants to try to use certificates on their YubiKey.
    - [Smartcard Certificate Picker Should Pop Up When Smartcard is Initially Plugged In](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1872882)
- Showing the user a dialog that lists the certificates available and prompts them to select one for authentication.
    - [Design Smartcard Certificate Picker](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1879582) 
- Prompting the user to enter the PIN that will unlock their YubiKey.
    - [Design PIN Prompt](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1872500)
    - [Get PIN Prompt to Pop Up When Certificate is Selected in Smartcard Cert Picker](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1879583)
- Handling cases when user wants to exit the flow.
    - [Handle Cancel Scenarios in Smartcard Cert Picker and PIN Prompt](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1879636)
- Handling the case when the user becomes locked out of their YubiKey due to too many incorrect PIN attempts.
    - [Design 'Too many attempts' PIN Dialog](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1914775)
- Ensuring that on-device CBA will remain functional.
    - [Main Success Scenario Works For On-Device Certificates](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1874347)
    - [On-Device Cert Picker should Pop Up When Smartcard is Not Plugged In](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1873478)  

This PR does NOT include the actual authentication with a private key, which is planned to be added next sprint cycle.
 
## Changes Made
`ClientCertAuthChallengeHandler` continues to handle the situation where a `ClientCertRequest` object is received from the server. It now also handles making connections to a plugged in YubiKey and reacting to user activity. The building and showing of dialogs are in separate classes, following the design of `HttpAuthDialog.java` (I decided not to use DialogFragments due to complications with bundles and threading, and I think that it's not very necessary because we are already controlling configuration changes, which is the main pro of using a DialogFragment, But if anyone has a strong opinion about this, please feel free to leave feedback). 

Resource files were also altered and added to match the required design made for the dialogs. I acknowledge that my implemented designs don't match perfectly (but I think I got pretty close :) ). Please see this [Figma link](https://www.figma.com/file/JhrzSYTAyKQ2nya3yGfOJe/AAD-sign-in-%2F-sign-up-features?node-id=702%3A23601) for the original designs.

Finally, the build.gradle file was updated to include compileOptions for Java 8, which is necessary in order to use YubiKit SDK.

## Result
When the user plugs in their YubiKey, gives permission via the Android dialog, and selects "sign in with certificate":
- The custom smartcard cert picker pop up with a listing of PIV certs on the key.
- User selects one and continues, which brings up another dialog that prompts them to enter a PIN.
- If they enter it in correctly, the dialog closes, and a Toast pops up yelling "Success!".
- If they don't enter it in correctly, the PIN dialog prompts them to reenter.
- If the user tries the maximum allotted times (usually 3). they see a different dialog that tells them they need to try again later.

If, at any point when a dialog is showing, the user decides to unplug their key, the dialog will be dismissed and the flow will be canceled.

## Video
I made a screen recording that was posted along with the link to this PR in the main Teams Code Review chat.

## Know Issues and Follow Ups
String resource translations will need to be done, which I believe we can get help for from the Authenticator team.

Sometimes, on the first try when starting up the webview and plugging in the YubiKey, the callback for `startUsbDiscovery` is not run, causing the local UsbDevice variable to remain null (which makes the on-device cert picker pop up instead). This happens sometimes for the first try, and it never happens on subsequent tries or if the Yubikey is plugged in before the webview is instantiated. I'll be doing some more investigation into how often this could occur and if there's any fixes that could be made.

A minor design issue I have noticed when testing with BrokerHost is that the radio buttons on the cert picker look a decade old. This doesn't affect the functionality, but it's something that I will probably look into to see if it's related to a broker library theme or a brokerHost theme.

## Testing
Once smartcard CBA goes into preview, I'll write some official test cases for broker and msal. 
Below are general ways to test the UI up until the actual authorization occurs.

**Beginning of all scenarios:**
1. Download AzureSample.
2. Click "Sign In".
3. Type the upn associated with the cert on key (or basically any other upn that brings up the "sign in with a certificate" link on the next page). Click "Next".
4. Plug in the YubiKey.
5. When the permission dialog pops up, click "OK".
6. Click "Sign in with a certificate". The custom cert picker should pop up with a listing of the certs currently held on the YubiKey.

**Success Scenario:**
7. Select any of the certs and click "Continue". The pin dialog should pop up.
8. Enter the YubiKey's pin and click "Unlock". The dialog should dismiss. A toast with "Success!" should pop up, and it is expected to see an error page.

**Incorrect Pin Scenario:**
7. Select any of the certs and click "Continue". The pin dialog should pop up.
8. Enter an incorrect YubiKey pin and click "Unlock". The EditText should clear and should have a red message underneath that says "The PIN you entered was incorrect".
9. Enter the YubiKey's pin and click "Unlock". The dialog should dismiss. A toast with "Success!" should pop up, and it is expected to see an error page.

**Max Pin Attempts Allowed Scenario:**
7. Select any of the certs and click "Continue". The pin dialog should pop up.
8. Enter an incorrect YubiKey pin and click "Unlock". The EditText should clear and should have a red message underneath that says "The PIN you entered was incorrect". 
9. Repeat step 8 two more times. The pin dialog should be replaced with another dialog that states "Too many failed attempts, try again later."
10. Click "Ok" on the dialog and it should dismiss.
11. Click the back button.
12. Retype the upn and click "next".
13. Click "Sign in with a certificate". The same "Too many failed attempts" dialog should pop up, and the error page should be seen.

**Early YubiKey Unplugging Scenario:**
7. Unplug the YubiKey. The cert picker dialog should dismiss and the error page should be shown.
8. Replug the YubiKey and click "Ok" when the permission prompt comes up.
9. Click the back button.
10. Retype the upn and click "next".
11. Click "Sign in with a certificate". The custom cert picker should pop up with a listing of the certs currently held on the YubiKey.
12. Select any cert and click "Next". The Pin dialog should pop up.
13. Unplug the YubiKey. The pin dialog should dismiss and the error page should be shown. 
14. Repeat steps 8-11.
15. Follow the "Max Pin Attempts Allowed Scenario", and when the error dialog shows up, unplug the YubiKey. The dialog should dismiss.

**Switch To On-Device Certs Scenario:**
7. Click "Cancel". The dialog should dismiss and the error page should be shown.
8. Unplug the YubiKey.
9. Click the back button.
10. Retype the upn and click "next".
11.  Click "Sign in with a certificate". The default Android on-device cert picker should pop up.


All of the scenarios above should be repeated with brokerHost installed as well.

melissaahn committed:
in progress, added logic for pin prompt

melissaahn committed:
restructed and added better comments

melissaahn committed:
pin prompt matches designers design

melissaahn committed:
works, but need to separate dialogs into classes

melissaahn committed:
created separate classes for dialogs

melissaahn committed:
added abstract SmartcardDialog class

melissaahn committed:
Merge remote-tracking branch 'origin/dev' into melissaahn/certpicker

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,311 @@
-        return handleOnDeviceCertAuth(request);
+        final String methodTag = TAG + ":processChallenge";
+        // Check if a YubiKey device has been discovered and has permission to be connected.
+        if (mDevice != null) {
```
(btw, if your concern is that nested methods are harder to read - you can utilize ResultFuture).

melissaahn replied:
If this is concerning the `if` statement, in which way could I use ResultFuture instead of synchronized here? The part that complicates it for me is the `else` statement, as if there isn't a device plugged in, we want to show the other on-device certs instead.

melissaahn replied:
Update: Tried using ResultFuture to wait for results from the cert picker (this implementation would have the overall same functionality, but I wouldn't have to separate the listeners into separate methods. This would be cleaner.) However, I noticed when I unplugged while the cert picker was showing, the thread was blocked by resultFuture.get(), despite me setting resultFuture.setResult(null) in the mDevice.setOnClosed() callback. This confuses me a bit, since I thought the setOnClosed() callback was running on a separate thread, but it seems that I may need to experiment more with ResultFuture if we want to take that route. 
EDIT: actually, it's probably getting blocked because I'm using synchronized on mDevice there... I'll test without that statement.
EDIT2: taking out synchronized everywhere doesn't make a difference. Will put a comment in the place where I wanted to put the ResultFuture and explain what happened.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,311 @@
+                        certPickerDialog.show();
+
+                    } catch(IOException e) {
+                        Logger.errorPII(methodTag, "IOException", e);
```
What would be the user experience if we run into any of these exceptions?

melissaahn replied:
The ClientCertRequest object should have 'cancel' called on it in these cases... which will cancel the flow and show the user an error message. I'll add these lines in.

rpdome replied:
Should we display an error message to the user?

melissaahn replied:
Would it be helpful for the user to see the error message? I was thinking that in these types of exceptions, it would be more helpful for the developers.. unless it would also be more helpful for us to do more than just error log.

rpdome replied:
@melissaahn probably. I would be confused if I plug yubikey in and i'm seeing the same error as if I don't plug it in.

melissaahn replied:
I'll look into creating an error dialog for these cases for next PR

rpdome replied:
let's add TODO comment in these blocks as a reminder

```Java
// TODO: XXXXXXXX
```

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,311 @@
                     @Override
                     public void run() {
                         Logger.info(TAG, "A YubiKey device was disconnected");
+                        mDevice = null;
```
What's the behavior if the customer unplugs the device in the middle of the flow (let's say they are picking the cert)

is it  
1. We cancel everything - display an error dialog, then cancel the request.
or
2. We let them proceed anyway, then after they click "Continue", then we display the error message and cancel the request - if the device does NOT match.
    > If this is the case, do we refresh the page if a different device is plugged in during the flow?

rpdome replied:
To simplify the logic, what I would suggest is to go with 1

1. we load the device at the beginning of processChallenge.
2. we hook an object which allows callback registration, which will be triggered whenever the user unplugs the device
3. we register an error "you've unplugged the device" dialog into that callback object.
4. we register request.cancel into that callback object
5. whenever we display a dialog, we register a dismiss method into that callback object.

(we'll need to run this through vimala)

melissaahn replied:
The current implementation is more like number 2, although inconsistent with error handling. I agree though that number 1 is simpler, and I like that it encourages users to keep their smartcard plugged in for the entire authentication process. 
I'll ask Vimala regarding an "unplugged device" dialog... although I think it could potentially not be necessary if the eSTS error message can give more specific info to the user as to how to properly plug in their smartcard (I'll check to see if this is still happening by GA)

rpdome replied:
Alternatively, you could move the cert loader logic into the cert picker dialog, and refreshes upon plugging/unplugging.

melissaahn replied:
Messaged Vimala and she said the new server error message should take care of the early unplugging situation (no extra dialog). 
The cert picker refreshing is an interesting idea. This would mean that in the unplugging callback, the current dialog would need to be distinguished so that the current dialog isn't dismissed if it's the cert picker. And there needs to be some similar logic for the plugin callback, too. 
I think I'm still leaning towards just canceling the flow if unplugged, partially because I don't think it will be too difficult for users to just click the back button and retype their upn to restart...  what do you think?

rpdome replied:
how would the server know if the user unplugged - if all we do is calling cancel() on the challenge object? Could it tell a difference from the scenario where the challenge fails due to no cert, no yubikey?

melissaahn replied:
[ClientCertRequest](https://developer.android.com/reference/android/webkit/ClientCertRequest#proceed(java.security.PrivateKey,%20java.security.cert.X509Certificate[])): There are three methods that can be used to deal with the request: cancel, ignore, and proceed. Proceed is the only method that takes in parameters (a PrivateKey and a Certificate array). So there aren't a ton of options when it comes to sending the server details as to why the flow is being cancelled.
^As mentioned in another comment, I'll look into creating an error dialog similar to the max attempts one.

rpdome replied:
Let's add a todo comment here.

melissaahn replied:
Adding a TODO comment and would like to get input from the team on the importance of having a dialog pop up when yubikey is unplugged too early (and what the dialog should say).

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,311 @@
+            @Override
+            public void onClick(final YubiKitCertDetails certDetails) {
+                //Ensure that YubiKey has not been removed too early by checking if mDevice is equal to the initialDevice.
+                if (mDevice != initialDevice) {
```
If I unplug and plug the same device, would this still be vaild?

melissaahn replied:
nope, according to my testing. I think if we go with the "cancel flow if unplugged at all" option, this check won't really be needed anymore anyway

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,311 @@
     private static final String ACCEPTABLE_ISSUER = "CN=MS-Organization-Access";
     private Activity mActivity;
     private final YubiKitManager mYubiKitManager;
+    private UsbYubiKeyDevice mDevice;
```
should this be synchronized?

rpdome reviewed (commented):
left a few comments.

melissaahn committed:
added synchronization for mDevice

melissaahn committed:
added cancelCba callback to SmartcardDialog

melissaahn committed:
addressed more feedback

melissaahn commented:
made some changes, including:

- synchronizing mDevice
- Creating a new variable mCurrentDialog (and synchronizing that one, too)
- Added the interface CancelCbaCallback and the corresponding setter method, which allows the ClientCertAuthChallengeHandler to set a callback to cancel the ClientCertRequest when the Yubikey is unplugged.
- in mDevice.setOnClosed(), if mCurrentDialog is not null, calls the onCancelCba() SmartcardDialog method on it.
- Making sure that ClientCertRequest is being canceled when an exception is thrown due to a YubiKit method. 

(might edit this for readability later... running low on sleep)

melissaahn commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                            synchronized (smartcardDialogLock) {
+                                mCurrentDialog = certPickerDialog;
+                            }
+
```
I wanted to put a `ResultFuture<YubiKitCertDetails>` `get()` call here, and then set the `setResult(null)` in the `onCancel()` method above. However, it seems that the `get()` statement is also blocking the `setOnClosed()` runnable that is run whenever the yubikey is unplugged. The result is that nothing in the `setOnClosed()` runnable is ever run.

rpdome replied:
Not sure which line(s) are you referring to...

melissaahn replied:
Line 225

melissaahn committed:
addressing spotbugs

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                                Logger.infoPII(methodTag,  "User has reached the maximum failed attempts allowed.");
+                                return;
+                            }
+                            //Create ArrayList that contains cert details only pertinent to the cert picker.
```
Line 186-196 can be extracted into a function.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                            //Check if too many PIN attempts have been made.
+                            //If so, go directly to Error dialog.
+                            if (piv.getPinAttempts() == 0) {
+                                SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(mActivity);
```
Line 168-185 can be extracted into a function.

Whenever possible, always extract code snippet into a function for readability.

rpdome replied:
For example, this whole method can be broken down into..

```Java
public Void processChallenge(@NonNull final ClientCertRequest request) {
     try {
           processChallengeInternal(request);
     } catch (SomeException e) {
           // kill ongoing dialog.
           // display some error dialog if needed.
           request.cancel();
     } 
}


void processChallengeInternal(request) {
    final Device device = tryGetYubiKeyDevice();
    if (device != null) {
        processChallengeWithYubiKeyDevice(device, request);
    } else {
        processChallengeWithOnDeviceCert(request);
    }
}

void processChallengeWithYubiKeyDevice(device, request){
    UsbSmartCardConnection connection = getDeviceConnection(device, request);
    PivSession piv; // get piv from result

    if (piv.getPinAttempts() == 0){
        showTooManyAttemptsDialog(); //invoke a dialog with a custom message that can be set in this method.
        return;
    }

    final ArrayList<YubiKitCertDetails> certList = getCertDetails(piv);
    if (certList.isEmpty()){
        showNoCertFoundDialog();  //invoke a dialog with a custom message that can be set in this method.
        return;
    }

    showCertPickerDialog(certList, device);
}
```

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                                return;
+                            }
+                            //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+                            final SmartcardCertPickerDialog certPickerDialog = new SmartcardCertPickerDialog(certList, mActivity);
```
extract 203-218 into a separate function.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+        synchronized (deviceLock) {
+            if (mDevice != null) {
+                //A connection to the YubiKey needs to be made in order to read the certificates off it.
+                mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
```
can requestConnection be extracted into a helper synchronized function with ResultFuture?

i.e. 
```Java

Result<UsbSmartCardConnection, IOException> requestConnection() {
    ResultFuture<Result<UsbSmartCardConnection, IOException>> future = new ResultFuture<>();
    // requestconnection
    return future.get(timeout); // could be longer since we're already in the yubikey flow.
}

```

rpdome replied:
I assume this is not possible to do?

melissaahn replied:
I'm going to say that based on my initial testing, I will not be utilizing ResultFuture in this PR, but I do want to sync with you on this for the next PR to make sure I'm giving it a fair shot.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+    private void getAndPutCertDetailsInList(Slot slot, PivSession piv, ArrayList<YubiKitCertDetails> certList) {
+        final String methodTag = TAG + ":getCert";
+        try {
+            X509Certificate cert =  piv.getCertificate(slot);
```
nit: final

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
 
+    //Helper method created to handle reading certificates off YubiKey.
+    //This method should only be called within a callback upon creating a successful YubiKey device connection.
+    private void getAndPutCertDetailsInList(Slot slot, PivSession piv, ArrayList<YubiKitCertDetails> certList) {
```
nit: add final and nonnull/nullable to method args. here and everywhere.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
-                        Logger.info(TAG, "A YubiKey device was disconnected");
-                    }
-                });
+                synchronized (deviceLock) {
```
what would happen if we call closeUsbDiscovery() right away here?

melissaahn replied:
Then startUsbDiscovery needs to be called appropriately again when the connection closes.
Implementation right now has it so that usbDIscovery only occurs while in the fragment showing the WebView

rpdome replied:
Let me phrase it the other way....

assuming that we already get the device object, can I invoke closeUsbDiscovery() right away? Can I still use that device object to make a connection after that?

(was trying to figure out if we can simplify the "unplug and replug" scenario.

melissaahn replied:
~~Yes, within the `invoke` method, we could call `closeUsbDiscovery()` right away.
What were your thoughts in terms of simplifying that scenario?~~
Decided to test this, and actually, calling `closeUsbDiscovery()` automaticallly calls the callback for `setOnClosed()`... so this doesn't actually seem recommended.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
     public Void processChallenge(@NonNull final ClientCertRequest request) {
-        //final String methodTag = TAG + ":processChallenge";
+        final String methodTag = TAG + ":processChallenge";
+        // Check if a YubiKey device has been discovered and has permission to be connected.
```
Wondering if we should make mDevice a local variable.

If that can be done, then we don't have to worry about synchronization. We created one object and pass it all the way through the flow (as a final variable). The object will not be re-assigned.

Can we try extracting the mDevice initialization block in the constructor, and turn it into a synchronous function (with resultfuture, with a small timeout)?

rpdome replied:
(not sure if possible since i don't know if processChallenge() is executed on main thread, and startUsbDiscovery() requires main thread to be unblocked.)

melissaahn replied:
So it sounds like we're trying  to find a way to detect the `UsbYubiKeyDevice` at the beginning of process challenge.
It's actually possible to call `startUsbDiscovery()` within `processChallenge` now, as I learned through testing. (I think the lack of `configChanges` keyboard flags were the culprit when I attempted this at the beginning of development.) I attempted moving the `startUsbDiscovery()` block to the beginning of `processChallenge`, creating a `ResultFuture` variable, calling `get` on the variable after the `startUsbDiscovery()` block, and calling `setResults` within the `startUsbDiscovery()` callback. 
However, `ResultFuture.get()` called within processChallenge does seem to block the callback for `startUsbDiscovery`, so I think both are relying on the same thread.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                    @Override
+                    public void invoke(Result<UsbSmartCardConnection, IOException> value) {
+                        try {
+                            final SmartCardConnection c = value.getValue();
```
what if value returns an exception?

melissaahn replied:
The method itself seems to handle throwing an exception. When `getValue()` is called, an IOException is thrown if the value (the private variable `value` of the `Result`) null. 
https://github.com/Yubico/yubikit-android/blob/0f04eeb3c9c9fc06eca29883b4ef2fa90293ffad/core/src/main/java/com/yubico/yubikit/core/util/Result.java#L35
But let me know if I'm misunderstanding this.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                        @Override
+                        public void invoke(Result<UsbSmartCardConnection, IOException> value) {
```
same here. handle ioexception if returned.

melissaahn replied:
^replied to other comment referring to this same issue.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                                    mActivity.runOnUiThread(new Runnable() {
+                                        @Override
+                                        public void run() {
+                                            Toast.makeText(mActivity.getApplicationContext(), "Success!", Toast.LENGTH_LONG).show();
```
Is this going to be shipped? or is this for testing only?

melissaahn replied:
Testing - I marked the parts for testing from 322 - 337

rpdome replied:
Let's add a todo here to make sure we remove it before we ship.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,379 @@
+                                        // and the user will need to figure out a way to reset their key outside of our library.
+                                        request.cancel();
+                                        pinDialog.dismiss();
+                                        SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(mActivity);
```
probably can reuse the function extracted from 168-185

rpdome reviewed (commented):
Left some comments - mostly about readability.
(you'll need another reviewer btw).

melissaahn committed:
addressing most feedback

melissaahn committed:
update build.gradle with java 8

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
     public ClientCertAuthChallengeHandler(@NonNull final Activity activity) {
         mActivity = activity;
-        //Create and start YubiKitManager
+        synchronized (deviceLock) {
```
nit: I don't think we need 105-113.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+                    //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+                    showSmartcardCertPickerDialog(certList, request);
+
+                } catch(IOException e) {
```
nit: add final to all exceptions. here and everywhere.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+
+    //Helper method that returns a list of YubiKitCertDetails extracted from certificates located on the YubiKey.
+    //This method should only be called within a callback upon creating a successful YubiKey device connection.
+    private List<YubiKitCertDetails> getCertDetailsFromKey(@NonNull final PivSession piv) {
```
nit: method comments should look like this

```Java

    /**
     * Helper method that returns a list of YubiKitCertDetails extracted from certificates located on the YubiKey.
     * This method should only be called within a callback upon creating a successful YubiKey device connection.
     * 
     * @param piv [ARGS_DESCRIPTION]
     * @return [RETURNED_VALUE_DESCRIPTION]
     */
    private List<YubiKitCertDetails> getCertDetailsFromKey(@NonNull final PivSession piv) {

```

rpdome replied:
here and everywhere.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+    //This method should only be called within a callback upon creating a successful YubiKey device connection.
+    private List<YubiKitCertDetails> getCertDetailsFromKey(@NonNull final PivSession piv) {
+        //Create ArrayList that contains cert details only pertinent to the cert picker.
+        List<YubiKitCertDetails> certList = new ArrayList<>();
```
nit: final

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+    private List<YubiKitCertDetails> getCertDetailsFromKey(@NonNull final PivSession piv) {
+        //Create ArrayList that contains cert details only pertinent to the cert picker.
+        List<YubiKitCertDetails> certList = new ArrayList<>();
+        //We need to check every slot. If a slot is empty, an ApduException is called, in which case, we'll ignore it.
```
nit: this comment kind of confused me a bit... as ApduException is already handled in getAndPutCertDetailsInList().

I read this comment as "these method might throw ApduException - and we should just ignore it"

rpdome replied:
(If another type of exception is thrown - IOException/BadResponseException - should we still swallow it silently, or should we fail the whole operation?)

melissaahn replied:
TL;DR: any exception in `getAndPutCertDetailsInList` excluding FILE_NOT_FOUND ApduException will result in a cancelled flow and an error message. Altering both `getAndPutCertDetailsInLIst` and `GetCertDetailsFromKey` to throw the exceptions, and then `handleSmartcardAuth` can handle the exceptions and cancelling the flow.

Deleted the ignoring ApduException part of the comment, as it is already addressed in `getAndPutCertDetailsInList` and causes unnecessary confusion.
I looked into the `getCertificate` method of `PivSession`, and I think for IO and ApduException (excluding FILE_NOT_FOUND), we should definitely cancel the flow and show an error message.
I was going back and forth regarding BadResponseException, which seems to be thrown when the certificate is compressed or it cannot be parsed. On one hand, it wouldn't be great if the exception was thrown for a cert that would not be typically used for CBA, as the user could not authenticate even if their desired CBA cert was installed correctly. On the other hand, if the desired cert was the one throwing the BadResponseException, it would be super confusing to the user if they just didn't see that cert in the picker without any sort of error message. Since the latter scenario would probably be more likely to occur, I think the flow should be cancelled for BadResponseException too.
What I will do is alter both `getAndPutCertDetailsInLIst` and `GetCertDetailsFromKey` to throw the specified exceptions, and then `handleSmartcardAuth` can handle the exceptions and cancelling the flow.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+        } catch (BadResponseException e) {
+            Logger.errorPII(methodTag,"BadResponseException", e);
+        } catch (ApduException e) {
+            //If sw is 0x6a82 (27266), This is a FILE_NOT_FOUND error, which we should ignore since this means the slot is merely empty.
```
should probably log as verbose. 

In private preview, we can turn verbose logging on - so that we have as much details as possible if we need to debug.

(verbose is disabled by default).

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+        } catch (ApduException e) {
+            //If sw is 0x6a82 (27266), This is a FILE_NOT_FOUND error, which we should ignore since this means the slot is merely empty.
+            if (e.getSw() != 0x6a82) {
+                Logger.errorPII(methodTag,"ApduException", e);
```
errorPII means it contains any PII data. Is this the case?

(you will not get anything logged by xxxxPII if the customer opts out from PII collection).

melissaahn replied:
I'm not certain that any of the logging I'm adding actually has PII, so I changed all the logging to the non-PII version for now.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+            }
+        }
+    }
+    //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
```
nit: add a line between method.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
     public Void processChallenge(@NonNull final ClientCertRequest request) {
         //final String methodTag = TAG + ":processChallenge";
-        return handleOnDeviceCertAuth(request);
+        processChallengeInternal(request);
```
nit: if this is the only line being executed, then we can merge this with the internal method.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+            @Override
+            public void onClick() {
+                request.cancel();
+                mCurrentDialog = null;
```
probably don't need this.

(if we do, make sure we surround it with the lock)

melissaahn replied:
We do need to set `mCurrentDialog` to null because the negative button listener of the dialog will automatically dismiss the dialog. 
If the yubikey is unplugged after a user hits "cancel", the `setOnClosed` callback will check if mCurrentDialog is equal to null. If it isn't null, it will attempt to dismiss the dialog.  I believe attempting to dismiss a dialog twice like this might throw some sort of error.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+        //Show the dialog.
+        certPickerDialog.show();
+        //Set current dialog to certPickerDialog
+        synchronized (smartcardDialogLock) {
```
nit: should probably move show() into the block. here and everywhere.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,434 @@
+                            } catch(IOException e) {
+                                Logger.errorPII(methodTag, "IOException", e);
+                                request.cancel();
+                                synchronized (smartcardDialogLock) {
```
line 329-331 seems to be repeated = this could be extracted into a method.

rpdome commented:
Almost there :) 

Most of the comments in this review iteration is mostly nitpicking. 

Anything that you would like to address in future PR, please add the // TODO comment.

Feel free to ping me once you're ready for the next iteration of review.

melissaahn committed:
addressed more feedback and added doc comments

melissaahn committed:
added doc comments to dialog classes and minor edit to ChallengeHandler

shahzaibj commented on common/build.gradle:
```diff
@@ -60,6 +60,11 @@ android {
         }
     }
+    //Upgrading to Java 8 in order to use YubiKit SDK
+    compileOptions {
```
Let's build Authenticator app early with these changes to make sure it is working and so that we don't catch these things during release this time.

shahzaibj replied:
Or the manifest changes that is...whichever was causing problems with Authenticator

melissaahn replied:
We'll be testing with new builds of both Broker and MSAL, which should hopefully fix the manifest mismatch errors.

shahzaibj replied:
Yes. I'm suggesting that we do that ahead of time rather than waiting until the next month's release time to test this.

rpdome replied:
@shahzaibj  this wont be released next month. we'll cut a private preview build based on the current release, cherry-pick CBA change, and hand the apk directly to certain customers.

The actual release would be in broker August build. (which will be hitting prod in September)

shahzaibj replied:
Hmmm...so how do we plan to do the Broker July release? I believe we would release off of `dev` which has the manifest changes and is going to break Authenticator if consumed without consuming a corresponding newer version of MSAL that also consumes the same common. Sounds like we're going to revert again directly on the release branch just like we did for release 5.0.0?

rpdome replied:
@shahzaibj  this PR is not going into dev. This is going into  melissa's feature branch.

melissaahn replied:
@shahzaibj I'll probably just revert my changes on the release branch for the upcoming release. This would probably be easier than taking all the YubiKey changes out of the dev branch in each repo, right?

shahzaibj replied:
@rpdome Yes, understood. May be I shouldn't have left the feedback here on this PR. The feedback I was providing wasn't about this PR but more generally about all the code changes that have been made to support the YubyKit CBA work. Some of those changes are still in `dev` and @melissaahn reverted them on the `release/5.0.0` branch during the release last time because those changes were breaking Authenticator. @melissaahn is now re-adding those changes to `merge to release` branch on the PR @gitaumoses4 has to merge release release back to dev because otherwise merging that branch back to dev will remove the YubyKit changes from dev as well. From what you're saying it seems like those changes shouldn't have been `dev` to begin with?

Anyway, all I want is to not be in the situation in which we were in during the last release and had to revert changes on the release branch to unblock Broker release. As long as we can solve that problem somehow then we should be fine.

rpdome replied:
In that case... let's revert the breaking change in dev and cherry-pick the commits into the feature branch.

Anything that is not breaking and is not used should be fine in dev.

rpdome replied:
(We should have around 2 weeks from code complete date to resolve all the possible build issues while we're building the private preview build on the side)

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+    private UsbYubiKeyDevice mDevice;
+    private SmartcardDialog mCurrentDialog;
+    //Locks to help facilitate synchronization
+    private static final Object deviceLock = new Object();
+    private static final Object smartcardDialogLock = new Object();
```
These variable names can be prefixed with `s` to denote `static`

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            this.slot = slot;
+        }
+
+        public String getIssuerText() {
```
javadoc

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            return issuerText;
+        }
 
+        public String getSubjectText() {
```
javadoc

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            return subjectText;
+        }
+
+        public Slot getSlot() {
```
javadoc

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+    private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":handleSmartcardCertAuth";
+        //A connection to the YubiKey needs to be made in order to read the certificates off it.
+        mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
```
It might be a good idea to add null check here as well

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+        mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void invoke(@NonNull Result<UsbSmartCardConnection, IOException> value) {
```
```suggestion
            public void invoke(@NonNull final Result<UsbSmartCardConnection, IOException> value) {
```

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            public void invoke(@NonNull Result<UsbSmartCardConnection, IOException> value) {
+                try {
+                    final SmartCardConnection c = value.getValue();
+                    PivSession piv = new PivSession(c);
```
```suggestion
                    final PivSession piv = new PivSession(c);
```

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+                    final SmartCardConnection c = value.getValue();
+                    PivSession piv = new PivSession(c);
+                    //Check if too many PIN attempts have been made.
+                    if (piv.getPinAttempts() == 0) {
```
So this is remaining pin attempts basically

melissaahn replied:
yes... are you asking about the comment? I can change it to  `//Check if there are no PIN attempts remaining.`

shahzaibj replied:
I was referring to the method name i.e. `getPinAttempts`. I was hoping it would be named something like `getRemainingPinAttempts`. I believe this method is in YubyKit code so I don't think you can change this.

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+                    showSmartcardCertPickerDialog(certList, request);
+
+                } catch(final IOException e) {
+                    Logger.error(methodTag, "IOException", e);
```
Do we want to log `e.getMessage()` in the message here and below?

melissaahn replied:
oh, I incorrectly assumed that Logger would automatically log `e.getMessage()` as well. I'll replace the current message with `e.getMessage()` to all error log statements.

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onCancel() {
+                request.cancel();
```
Do we need to set `mCurrentDialog` to `null` here?

melissaahn replied:
I don't think so, as whenever we call `mCurrentDialog.onCancelCba()`, which calls the `CancelCbaCallback`, `mCurrentDialog` is immediately set to null afterwards

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            public void onClick(final String pin) {
+                synchronized (deviceLock) {
+                    // Need to request a SmartCardConnection in order to access certs on YubiKey.
+                    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
```
May want to check for `mDevice != null` here

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+                    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                        @Override
+                        public void invoke(@NonNull Result<UsbSmartCardConnection, IOException> value) {
```
```suggestion
                        public void invoke(@NonNull final Result<UsbSmartCardConnection, IOException> value) {
```

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+        try {
+            piv.verifyPin(pin);
+            //If pin is successfully verified, we will get the certificate and perform the rest of the logic for authentication.
+            X509Certificate cert = piv.getCertificate(certDetails.getSlot());
```
```suggestion
            final X509Certificate cert = piv.getCertificate(certDetails.getSlot());
```

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+            //TODO: Complete authentication using cert and YubiKit sdk.
+            //TODO: Delete test code below between START and END comments (should be replaced with actual authentication)
+            //NOTE: below is for testing. This would get replaced by the logic for actual authentication.
+            //TODO:START of code for testing
```
This is going to be replaced in next few days and NOT ship? Or rather is this testing code going to ship?

melissaahn replied:
replaced in next few days and NOT ship (it has already been replaced in the next PR I'm working on).

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+        } catch (final InvalidPinException e) {
+            // An incorrect Pin attempt.
+            // We need to retrieve the number of pin attempts before proceeding.
+            int pinAttemptsRemaining = piv.getPinAttempts();
```
```suggestion
            final int pinAttemptsRemaining = piv.getPinAttempts();
```

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+     */
+    private void showMaxAttemptsDialog(){
+        //final String methodTag = TAG + ":showMaxAttemptsDialog";
+        SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(mActivity);
```
```suggestion
        final SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(mActivity);
```

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,64 +22,490 @@
+     * Sets up and shows dialog when user has exceeded the max attempts allowed to enter their YubiKey PIN.
+     */
+    private void showMaxAttemptsDialog(){
+        //final String methodTag = TAG + ":showMaxAttemptsDialog";
```
Does this need to be removed?

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertPickerDialog.java:
```diff
@@ -0,0 +1,222 @@
+     * @param certList List of ClientCertAuthChallengeHandler.YubiKitCertDetails compiled from certificates on YubiKey.
+     * @param activity Host activity.
+     */
+    public SmartcardCertPickerDialog(List<ClientCertAuthChallengeHandler.YubiKitCertDetails> certList, Activity activity) {
```
```suggestion
    public SmartcardCertPickerDialog(@NonNull final List<ClientCertAuthChallengeHandler.YubiKitCertDetails> certList, @NonNull final Activity activity) {
```

melissaahn replied:
Will add `@NonNull final` to other methods within all SmartcardDialog classes

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertPickerDialog.java:
```diff
@@ -0,0 +1,222 @@
+                //Start building the dialog.
+                AlertDialog.Builder builder = new AlertDialog.Builder(mActivity, R.style.CertAlertDialogTheme)
+                        .setTitle(R.string.smartcard_cert_dialog_title)
+                        .setSingleChoiceItems(certAdapter, 0, null)
```
Can you add some comments here? What are all the params here?

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertPickerDialog.java:
```diff
@@ -0,0 +1,222 @@
+     * Sets listener for positive button.
+     * @param listener Implemented PositiveButtonListener.
+     */
+    public void setPositiveButtonListener(PositiveButtonListener listener) {
```
These listeners are needed to be re-assigned? and hence providing setters for these? Can these instead be supplied on object construction for this class?

shahzaibj replied:
The same question for all the listeners

melissaahn replied:
I think these can be supplied on object construction. I'll make these changes for all the listeners

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertPickerDialog.java:
```diff
@@ -0,0 +1,222 @@
+    /**
+     * Builds an AlertDialog that displays the details of the certificates in a single choice ListView and prompts the user to choose a certificate to proceed.
+     */
+    protected void createDialog() {
```
This is create and Show dialog? Or only create?

melissaahn replied:
Only create. `show()` is inherited from the `SmartcardDialog` class in `SmartcardCertPickerDialog`

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardDialog.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    /**
+     * Should build an Android Dialog object and set it to mDialog.
```
> set it to mDialog

Sounds like this should be happening there on line 42?

melissaahn replied:
I didn't want to call `createDialog()` on the last line in the `SmartcardDialog` constructor because there may be child classes with `createDialog()` implementations that rely on instance variables. `SmartcardCertPickerDialog` is one such class where the List of certificate details is needed before `createDialog` is called. But if `createDialog` is called in the parent constructor, the dialog will be created too early without the cert details list.

melissaahn committed:
addressed first part of new feedback

melissaahn committed:
address 2nd part of feedback

melissaahn commented:
Addressed current feedback. Let me know if I missed anything or didn't address something correctly.
Some notes:
- I found out that YubiKey PINs can actually [contain all types of characters](https://developers.yubico.com/yubikey-piv-manager/PIN_and_Management_Key.html), so I changed the PIN EditText to `textPassword` instead of `numberPassword`
- In the next PR, I think I'm going to modify the `SmartcardMaxFailedAttemptsDialog` to become a generic `ErrorDialog` that will take in string resources as constructor parameters, since all the error dialogs will visually look the same.

rpdome reviewed (approved):
:shipit:

shahzaibj reviewed (approved):
Thanks @melissaahn!
